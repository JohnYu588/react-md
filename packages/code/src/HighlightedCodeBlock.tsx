import { cnb } from "cnbuilder";
import { type HTMLAttributes, type ReactElement } from "react";
import { CodeBlock, type CodeBlockConfigurableProps } from "./CodeBlock.js";
import { type HighlightCode, type SupportedCodeLanguage } from "./types.js";

const LANGUAGE_REGEX = /language-([a-z]+)/;

export interface BaseHighlightedCodeBlockProps
  extends CodeBlockConfigurableProps {
  children: string;
  language?: SupportedCodeLanguage;
  className?: string;
  codeProps?: HTMLAttributes<HTMLElement>;
}

export interface HighlightedCodeBlockProps
  extends BaseHighlightedCodeBlockProps {
  highlightCode: HighlightCode;
}

export function HighlightedCodeBlock(
  props: HighlightedCodeBlockProps & { language: SupportedCodeLanguage }
): ReactElement;
export function HighlightedCodeBlock(
  props: HighlightedCodeBlockProps & { className: string }
): ReactElement;
export function HighlightedCodeBlock(
  props: HighlightedCodeBlockProps & {
    language: SupportedCodeLanguage;
    className: string;
  }
): ReactElement;
export function HighlightedCodeBlock(
  props: HighlightedCodeBlockProps
): ReactElement {
  const {
    children,
    codeProps,
    language,
    className: propClassName,
    highlightCode,
    ...remaining
  } = props;

  let lang: string | undefined = language;
  let className = propClassName;
  if (!lang && propClassName) {
    [, lang] = propClassName.match(LANGUAGE_REGEX) || [];
  }
  className ??= `language-${lang}`;

  if (!lang) {
    throw new Error(
      `HighlightedCodeBlock must provide a language by the \`language\` prop or \`className="language-${language}"\``
    );
  }

  return (
    <CodeBlock {...remaining} className={className}>
      <code
        {...codeProps}
        className={cnb(className, codeProps?.className)}
        dangerouslySetInnerHTML={{
          __html: highlightCode({ code: children, lang }),
        }}
      />
    </CodeBlock>
  );
}
