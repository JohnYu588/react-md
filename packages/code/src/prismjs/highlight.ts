import Prism from "prismjs";
import "prismjs/components/prism-bash";
import "prismjs/components/prism-css";
import "prismjs/components/prism-css-extras";
import "prismjs/components/prism-diff";
import "prismjs/components/prism-git";
import "prismjs/components/prism-javascript";
import "prismjs/components/prism-json";
import "prismjs/components/prism-jsx";
import "prismjs/components/prism-markdown";
import "prismjs/components/prism-markup";
import "prismjs/components/prism-properties";
import "prismjs/components/prism-scss";
import "prismjs/components/prism-tsx";
import "prismjs/components/prism-typescript";

import { type HighlightCode } from "../types.js";

Prism.manual = true;

export const highlightCode: HighlightCode = (options): string => {
  const { code, lang } = options;
  return Prism.highlight(code, Prism.languages[lang], lang);
};
