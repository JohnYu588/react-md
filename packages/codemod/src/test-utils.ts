// This is a mini port of:
// https://github.com/facebook/jscodeshift/blob/b07b6c9c46a3198d13b21a004a06409e8c347fc3/src/testUtils.js#L123
// to be used for testing with ESM only
import { describe, expect, it } from "@jest/globals";
import jscodeshift, {
  type Options,
  type Parser,
  type Transform,
} from "jscodeshift";
import { type TestOptions } from "jscodeshift/src/testUtils.js";
import { readFile } from "node:fs/promises";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

interface TransformModule {
  default: Transform;
  parser?: TestOptions["parser"];
}

function getExtension(parser?: string | Parser): string {
  if (parser === "ts" || parser === "tsx") {
    return parser;
  }

  return "js";
}

export async function applyTransform(
  transformer: TransformModule | Transform,
  options: Options | null | undefined,
  input: Parameters<Transform>[0],
  testOptions?: TestOptions
): Promise<string> {
  const parser =
    testOptions?.parser ??
    (("parser" in transformer && transformer.parser) || "tsx");
  const transform =
    typeof transformer === "function" ? transformer : transformer.default;
  const j = jscodeshift.withParser(parser);
  const output = await Promise.resolve(
    transform(
      input,
      {
        jscodeshift: j,
        j: jscodeshift,
        stats: () => {},
        report: () => {},
      },
      options || {}
    )
  );

  return (output || "").trim();
}

export function defineTest(
  importMetaUrl: string,
  transformName: string,
  options?: Options | null,
  testFilePrefix = "",
  testOptions?: TestOptions
): void {
  const suffix = testFilePrefix ? `using "${testFilePrefix}" data` : "";
  const testName = "transforms correctly " + suffix;
  const dirName = dirname(fileURLToPath(importMetaUrl));

  describe(transformName, () => {
    it(testName, async () => {
      const transformer: TransformModule = await import(
        join(dirName, "..", transformName)
      );
      const parser = testOptions?.parser ?? transformer.parser ?? "tsx";
      const extension = getExtension(parser);

      const fixtureDir = join(dirName, "..", "__testfixtures__");
      const inputPath = join(
        fixtureDir,
        testFilePrefix + `.input.${extension}`
      );
      const outputPath = inputPath.replace(".input.", ".output.");

      const source = await readFile(inputPath, "utf8");
      const expected = await readFile(outputPath, "utf8");

      const output = await applyTransform(
        transformer,
        options,
        { source, path: inputPath },
        testOptions
      );

      expect(output).toEqual(expected.trim());
    });
  });
}
