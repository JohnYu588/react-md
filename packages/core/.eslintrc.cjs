module.exports = {
  root: true,
  extends: ["@mlaursen/eslint-config"],
  parserOptions: {
    project: "./tsconfig.json",
    tsconfigRootDir: __dirname,
    ecmaVersion: 2018,
    sourceType: "module",
  },
  rules: {
    "@typescript-eslint/consistent-type-imports": ["error"],
  },
  overrides: [
    {
      files: ["src/**/index.ts"],
      rules: {
        // I don't know how to get @module to work but it's supported by typedoc
        "tsdoc/syntax": 0,
      },
    },
  ],
};
