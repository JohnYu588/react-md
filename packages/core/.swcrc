{
  "$schema": "https://json.schemastore.org/swcrc.json",
  "exclude": ["__tests__", "src/test-utils/drag.ts"],
  "jsc": {
    "parser": {
      "syntax": "typescript",
      "tsx": true
    },
    "transform": {
      "react": {
        "runtime": "automatic"
      }
    },
    "target": "es2021"
  },
  "sourceMaps": true
}
