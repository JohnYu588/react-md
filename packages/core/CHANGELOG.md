# @react-md/core

## 1.0.0-next.4

### Patch Changes

- Fixed Sass variable issues and other CSS tweaks.

## 1.0.0-next.3

### Patch Changes

- The next v6 alpha release.

  # Changes

  ## AppBar

  - Updated `AppBar` to use the `Box` component and additional `cssUtils`
  - Added a `$nav-keyline` variable for the `AppBarTitle` offset

  ## Avatar

  - Fixed the `color` prop definition

  ## Badge

  - Simplified the offset variables

  ## Button

  - Added a `TooltippedButton` component

  ## Card

  - Added a `ClickableCard` component
  - Fixed the `CardSubtitle` color

  ## Dialog

  - Fixed the background colors

  ## Divider

  - Removed the `vertical-sizing` variable and updated the vertical divider to update the same way as a horizontal divider

  ## Expansion Panel

  - Renamed `preventAllClosed` to `preventAllCollapsed` to better match naming schemes

  ## Form

  - Updated `MenuItemInputToggle` to use `multiline` instead of `threeLines`
  - Fixed the `Select` focus behavior when rendering within a `Sheet`

  ## Icon

  - Fixed the styling for the `IconRotator` by setting the default rotate variables
  - Fixed the styling for the `IconSpacing` component when using `above` or `below`

  ## List

  - Renamed `threeLines` to `multiline` and updated the CSS variables to match
  - Fixed some styling by correctly passing all props down
  - Added the ability to disable the spacing for the left addons

  ## Portal

  - Updated `PortalContainerProvider` to support a `RefObject`

  ## Table

  - Fixed the `TableCheckbox` and `TableRadio` components to be client components instead of server components

  ## Tabs

  - Added support for rendering vertically
  - Added support to render a tab as a link instead of a button
  - Renamed the `--rmd-tab-width` to `--rmd-tab-size`

  ## Tooltip

  - Fixed some type definition issues for event handlers

  ## Transition

  - Updated `SkeletonPlaceholder` to `forwardRef`
  - Updated `SkeletonPlaceholder` and `useSkeletonPlaceholder` to support `animation-delay`
  - Added `randomSkeletonPlaceholder` util to generate SSR-only skeletons
  - Fixed the naming of the Sass variables since they weren't prefixed by `transition-`

  ## Tree

  - Fixed the type definitions for the `DefaultTreeItemRenderer` for the `data` prop after switching to `RenderRecursively`
  - Fixed the `Tree` components to start the `--rmd-tree-depth` at `0` instead of `-1`
  - Removed the `--padded` style since it's no longer needed

  ## Typography

  - Updated the base typography to default to `overflow-wrap: break-word`
  - Fixed the `TextContainer` styling to include the padding in the line length

  ## Other

  - Uses `:has` selectors and `:focus-visible` since all major browsers support them now
  - Uses `@media(hover: hover) and (pointer: fine)` for hover states instead of `.mouse-mode`
  - Refactored some draggable utils and fixed bugs around dragging
  - Added a `useElementSize` wrapper for the `useResizeObserver` hook
  - Fixed `useElementSize` to trigger immediately on mount
  - Updated `useWindowSize` to support SSR values

## 1.0.0-next.2

### Major Changes

- Updated custom properties to require a defined value instead of always providing a fallback. This decreases the bundle size by a good amount.

  Also updated the `Snackbar` so that it no longer has a `role="status"` to fix accessibility. Instead, each toast will have `role="status"` or `role="alert"`.

## 1.0.0-next.1

### Major Changes

- 117350ed79: The next preview of v6.0.0.

  Breaking Changes:

  - An accessible label is now required for `CircularProgress` and `LinearProgress`
    - Updated `AsyncButton` to pass props to the progress component and other
      accessibility teaks
    - Updated `CircularProgressSuspense` to default the `aria-label` to `"Loading"`
  - Removed the `VerticalDivider` component
  - Rename `hoverTime`/`leaveTime` to `hoverTimeout`/`leaveTimeout` for hover mode
  - Removed the `forceSize`/`forceFontSize` props from the `FontIcon`
  - Removed the `useOverflowTooltip` and instead integrate it into the
    `useTooltip` hook. It's cheaper to check on the start event instead of having
    a resize observer always active
  - Removed the `LinkProvider` since it was only used for the `Tree` component and
    it's easier to pass the custom link component down

  Features:

  - Added `TRANSITION_CONFIG` so transitions can be disabled
    - Updated the `@react-md/test-utils/jest-setup` to always disable transitions
      in tests
  - Added `MATERIAL_CONFIG` to override defaults instead of using context
  - Adding support for including `data-testid` in all HTMLAttributes by including
    `@react-md/test-utils/data-testid` somewhere (normally `setupTests.ts`)
  - Added `theme` to `Avatar`
  - Update `Dialog` to be an `alertdialog` by default when `modal` is enabled and
    added `aria-modal` when `modal` is enabled
  - Add additional props to `Slider` to customize the tooltip
  - Updated all table styling utility functions to be in separate files so they
    aren't forced as client code
  - Added `iconRotatorProps` for the `TableCell`
  - Updated `isStickyActive` to provide the `isInTableContainer` boolean and
    export `isTableFooterStickyActive` default function
  - Add support for `stickyActiveClassName`
  - Added support for `TableRadio` component
  - Updated `DropdownMenu` to support controlling the visibility

  Bug Fixes:

  - Improved performance for multiple transitions by no longer causing layout
    shifts
  - Add some missing custom properties to inline styles for Typescript
  - Update `Tooltip` to default to temporary again. There's no need to create 1000
    nodes in large apps on startup
  - Fixed styles for `MenuItemCheckbox`, `MenuItemRadio`, and `MenuItemSwitch`
  - Fix `Slider` marks when the min value is not `0`
  - Fixed the default anchor for the `Menu` component when using a floating action
    button
  - Update `Tooltip` to use `textOverflow` instead of `disableLineWrap`
  - Update `Switch` to no longer be a client component
  - Update `Switch` to support rendering components within the ball

## 1.0.0-next.0

### Major Changes

- Initial v6 release test
