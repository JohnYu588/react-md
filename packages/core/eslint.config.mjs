// @ts-check
import { config, configs, gitignore } from "@mlaursen/eslint-config";
import { join } from "node:path";

// eslint-disable-next-line no-undef
const strict = process.env.STRICT_TYPING === "true";

const frontend = strict
  ? configs.frontendTypeChecking(import.meta.dirname)
  : configs.frontend;

export default config(
  gitignore(join(import.meta.url, "..", "..")),
  ...frontend
);
