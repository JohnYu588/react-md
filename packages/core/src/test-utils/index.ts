export * from "@testing-library/react";
export * from "@testing-library/user-event";
export * from "./mocks/match-media.js";
export * from "./render.js";
export * from "./utils/resize-observer.js";
