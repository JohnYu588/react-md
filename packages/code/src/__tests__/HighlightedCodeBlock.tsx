import { describe, expect, it } from "@jest/globals";
import { render, screen } from "@react-md/core/test-utils";

import { HighlightedCodeBlock } from "../HighlightedCodeBlock.js";
import { highlightCode } from "../prismjs/highlight.js";

describe("HighlightedCodeBlock", () => {
  it("should auto highlight children", () => {
    render(
      <HighlightedCodeBlock
        containerProps={{ "data-testid": "container" }}
        language="ts"
        highlightCode={highlightCode}
      >
        const x = 3;
      </HighlightedCodeBlock>
    );

    const container = screen.getByTestId("container");
    expect(container).toMatchSnapshot();
  });
});
