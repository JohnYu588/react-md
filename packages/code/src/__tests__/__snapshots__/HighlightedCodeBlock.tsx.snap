// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HighlightedCodeBlock should auto highlight children 1`] = `
<div
  class="code-block rmd-text-primary-color rmd-dark-surface"
  data-testid="container"
>
  <div
    class="code-block__scroll-container"
  >
    <div
      class="code-block__pre-container"
    >
      <pre
        class="language-ts code-block__pre"
      >
        <code
          class="language-ts"
        >
          <span
            class="token keyword"
          >
            const
          </span>
           x 
          <span
            class="token operator"
          >
            =
          </span>
           
          <span
            class="token number"
          >
            3
          </span>
          <span
            class="token punctuation"
          >
            ;
          </span>
        </code>
      </pre>
    </div>
  </div>
</div>
`;
